src/main/scala/ie/zalando/fabric/gateway/service/IngressDerivationChain.scala

```diff
@@ -90,8 +90,33 @@ class IngressDerivationChain(stackSetOperations: StackSetOperations)(implicit ma
       val finalRoutes = gateway.corsConfig.fold(skipperRoutes) { corsConfig =>
         withCors(gateway, meta.name, corsConfig, skipperRoutes)
       }
-      combineBackendsAndRoutes(backends, finalRoutes, meta)
+      val ingressDefinitions = combineBackendsAndRoutes(backends, finalRoutes, meta)
+
+      gateway.serviceProvider match {
+        case StackSetProvidedServices(_, _) if versionedHostsEnabled => appendServiceSpecificHosts(ingressDefinitions)
+        case _ => ingressDefinitions
+      }
+    }
+  }
+
+  def appendServiceSpecificHosts(ingresses: List[IngressDefinition]): List[IngressDefinition] = {
+    val serviceSpecificIngresses = for {
+      ingress <- ingresses
+      hostMapping <- ingress.hostMappings
+      splitHost = hostMapping.host.split('.')
+      if splitHost.length == 4 && splitHost(2) == "zalan" && splitHost(3) == "do"
+      service <- hostMapping.services
+      versionedHost = s"${service.name}.${splitHost.drop(1).mkString(".")}"
+    } yield {
+      val name = s"${ingress.metadata.name}-${service.name}"
+      val annotations  = ingress.metadata.routeDefinition.additionalAnnotations.filterKeys(_ != "zalando.org/backend-weights")
+      val routeDefinition = ingress.metadata.routeDefinition.copy(additionalAnnotations = annotations)
+      val metadata = ingress.metadata.copy(name = name, routeDefinition = routeDefinition)
+
+      ingress.copy(metadata = metadata, hostMappings = Set(IngressBackend(versionedHost, Set(service))))
     }
+
+    ingresses ++ serviceSpecificIngresses
```

Comment:
What happens with the default routes? Are they covered by this?

Reply:
They are yeah

It's like the following:
```
  owned_ingress_names:
  - managed-by-stackset-1-reject-http-route
  - managed-by-stackset-1-default-404-route-ss-stack-v2
  - managed-by-stackset-1-default-404-route-ss-stack-v1
  - managed-by-stackset-1-get-api-all-ss-stack-v2
  - managed-by-stackset-1-default-404-route
  - managed-by-stackset-1-reject-http-route-ss-stack-v2
  - managed-by-stackset-1-get-api-all-ss-stack-v1
  - managed-by-stackset-1-get-api-all
  - managed-by-stackset-1-reject-http-route-ss-stack-v1

```

Reply:
What's the gateway name on Playground? I'll have a look...

Reply:
`managed-by-stackset-1`